fs-storage/src/monoid.rs

```diff
@@ -0,0 +1,46 @@
+// Currently, we have three structures: Tags (HashSet), Properties (HashSet), Score (int).
+// In fact, HashSet already implements a union function,
+// so only a special function for integers is needed.
+// CRDTs can be considered later when we need to add structures that require
+// more powerful combine semantics.
+
+// Trait defining a Monoid, which represents a mathematical structure with an identity element and an associative binary operation.
+pub trait Monoid<V> {
+    // Returns the neutral element of the monoid.
+    fn neutral() -> V;
+
+    // Combines two elements of the monoid into a single element.
+    fn combine(a: &V, b: &V) -> V;
+
+    // Combines multiple elements of the monoid into a single element.
+    // Default implementation uses `neutral()` as the initial accumulator and `combine()` for folding.
+    fn combine_all<I: IntoIterator<Item = V>>(values: I) -> V {
+        values
+            .into_iter()
+            .fold(Self::neutral(), |acc, val| Self::combine(&acc, &val))
+    }
+}
+
+impl Monoid<i32> for i32 {
+    fn neutral() -> i32 {
+        0
+    }
+
+    fn combine(a: &i32, b: &i32) -> i32 {
+        if a > b {
+            *a
+        } else {
+            *b
+        }
+    }
+}
+
+impl Monoid<String> for String {
+    fn neutral() -> String {
+        "".to_string()
+    }
+
+    fn combine(a: &String, b: &String) -> String {
```

Comment:
Also this way of combining will be incredibly inefficient for Strings because it'll reallocate on each combine. Typically practical monoid implementations also include a `combine_all` that takes a an iterator of values and combines them in an efficient manner. Like Python's join operator that takes an iterator of strings instead of `+` each string individually.

This is not relevant immediately but something to keep in mind when monoids are used to combine over iterators (aka folds).